File: java/src/com/sandy/fda/utils/FDACrypto.java
package com.sandy.fda.utils;

import javax.crypto.Cipher;
import javax.crypto.Mac;
import javax.crypto.spec.GCMParameterSpec;
import javax.crypto.spec.SecretKeySpec;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.security.SecureRandom;
import java.util.Base64;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Machine-bound AES-256-GCM encryption/decryption, matching FDACrypto.cpp
 * exactly. Both sides must produce byte-identical keys given the same
 * MachineGuid — see the STATIC_SECRET comment below.
 *
 * Key = HMAC-SHA256(key = STATIC_SECRET, message = MachineGuid)
 */
public final class FDACrypto {

    private static final byte[] STATIC_SECRET = {
        0x4A, 0x2F, (byte)0x8C, 0x1D, 0x6E, (byte)0x93, 0x5B, 0x7A,
        (byte)0xC4, 0x0E, 0x2B, (byte)0xF1, (byte)0x88, 0x3D, (byte)0x9A, 0x56,
        0x71, (byte)0xE0, 0x4C, (byte)0xA8, 0x2D, 0x67, (byte)0xB3, (byte)0xF9,
        0x1A, 0x5E, (byte)0xC2, (byte)0x84, 0x39, (byte)0xD6, 0x0F, 0x7B
    };

    private static final int GCM_IV_LENGTH = 12;
    private static final int GCM_TAG_LENGTH_BITS = 128; // 16 bytes

    private FDACrypto() {}

    public static String encrypt(String plaintext) {
        try {
            SecretKeySpec key = deriveKey();
            if (key == null) return "";

            byte[] iv = new byte[GCM_IV_LENGTH];
            new SecureRandom().nextBytes(iv);

            Cipher cipher = Cipher.getInstance("AES/GCM/NoPadding");
            GCMParameterSpec spec = new GCMParameterSpec(GCM_TAG_LENGTH_BITS, iv);
            cipher.init(Cipher.ENCRYPT_MODE, key, spec);

            byte[] ciphertextAndTag = cipher.doFinal(plaintext.getBytes("UTF-8"));

            byte[] combined = new byte[iv.length + ciphertextAndTag.length];
            System.arraycopy(iv, 0, combined, 0, iv.length);
            System.arraycopy(ciphertextAndTag, 0, combined, iv.length, ciphertextAndTag.length);

            return Base64.getEncoder().encodeToString(combined);
        } catch (Exception e) {
            throw new IllegalStateException("[CRYPTO] encrypt failed." , e);
        }
    }

    public static String decrypt(String base64Ciphertext) {
        try {
            SecretKeySpec key = deriveKey();
            if (key == null) return "";

            byte[] combined = Base64.getDecoder().decode(base64Ciphertext);
            if (combined.length < GCM_IV_LENGTH + 16) return ""; 

            byte[] iv = new byte[GCM_IV_LENGTH];
            System.arraycopy(combined, 0, iv, 0, GCM_IV_LENGTH);

            byte[] ciphertextAndTag = new byte[combined.length - GCM_IV_LENGTH];
            System.arraycopy(combined, GCM_IV_LENGTH, ciphertextAndTag, 0, ciphertextAndTag.length);

            Cipher cipher = Cipher.getInstance("AES/GCM/NoPadding");
            GCMParameterSpec spec = new GCMParameterSpec(GCM_TAG_LENGTH_BITS, iv);
            cipher.init(Cipher.DECRYPT_MODE, key, spec);

            byte[] plaintext = cipher.doFinal(ciphertextAndTag);
            return new String(plaintext, "UTF-8");
        } catch (Exception e) {
            // Includes AEADBadTagException — tampered or corrupted value
            throw new IllegalStateException("[CRYPTO] decrypt failed — value may be corrupted or tampered with. " + e);
        }
    }

    private static SecretKeySpec deriveKey() throws Exception{
        try {
            String machineGuid = getMachineGuid();
            if (machineGuid == null || machineGuid.isEmpty()) return null;

            Mac hmac = Mac.getInstance("HmacSHA256");
            hmac.init(new SecretKeySpec(STATIC_SECRET, "HmacSHA256"));
            byte[] keyBytes = hmac.doFinal(machineGuid.getBytes("UTF-8"));

            return new SecretKeySpec(keyBytes, "AES");
        } catch (Exception e) {
            throw new IllegalStateException("[CRYPTO] deriveKey failed.",e);
        }
    }

    private static String getMachineGuid() {
        try {
            ProcessBuilder pb = new ProcessBuilder(
                "reg", "query",
                "HKLM\\SOFTWARE\\Microsoft\\Cryptography",
                "/v", "MachineGuid"
            );
            pb.redirectErrorStream(true);
            Process proc = pb.start();

            StringBuilder output = new StringBuilder();
            try (BufferedReader reader = new BufferedReader(new InputStreamReader(proc.getInputStream()))) {
                String line;
                while ((line = reader.readLine()) != null) {
                    output.append(line).append('\n');
                }
            }
            proc.waitFor();

            // Expected line format: "    MachineGuid    REG_SZ    <guid>"
            Pattern pattern = Pattern.compile("MachineGuid\\s+REG_SZ\\s+(\\S+)");
            Matcher matcher = pattern.matcher(output.toString());
            if (matcher.find()) {
                return matcher.group(1);
            }
            throw new IllegalStateException("[CRYPTO] MachineGuid not found in reg query output");
        } catch (Exception e) {
            throw new IllegalStateException("[CRYPTO] getMachineGuid failed.",e);
        }
    }
}

File: cpp/src/Utils/FDACrypto.cpp
#define WIN32_LEAN_AND_MEAN
#include "FDACrypto.h"
#include "Logger.h"
#include <Windows.h>
#include <bcrypt.h>
#include <vector>
#include <string>

#pragma comment(lib, "bcrypt.lib")

// -------------------------------------------------------
// Static secret — MUST match FDAConstants.java's STATIC_SECRET exactly,
// byte for byte. This alone is not the encryption key; it's combined
// with the machine's MachineGuid via HMAC-SHA256 (see deriveKey()).
// -------------------------------------------------------
static const unsigned char STATIC_SECRET[32] = {
    0x4A, 0x2F, 0x8C, 0x1D, 0x6E, 0x93, 0x5B, 0x7A,
    0xC4, 0x0E, 0x2B, 0xF1, 0x88, 0x3D, 0x9A, 0x56,
    0x71, 0xE0, 0x4C, 0xA8, 0x2D, 0x67, 0xB3, 0xF9,
    0x1A, 0x5E, 0xC2, 0x84, 0x39, 0xD6, 0x0F, 0x7B
};
// NOTE: replace these with your own randomly generated 32 bytes before
// shipping — the values above are placeholders for illustration only.
// Generate real ones once (e.g. via `openssl rand -hex 32` split into
// byte literals) and hardcode identically on both sides.

// -------------------------------------------------------
// Base64 helpers (Windows CryptStringToBinary/CryptBinaryToString —
// already part of the OS, no new dependency)
// -------------------------------------------------------
#include <wincrypt.h>
#pragma comment(lib, "crypt32.lib")

static std::string base64Encode(const std::vector<unsigned char>& data)
{
    DWORD outLen = 0;
    if (!CryptBinaryToStringA(data.data(), static_cast<DWORD>(data.size()),
        CRYPT_STRING_BASE64 | CRYPT_STRING_NOCRLF, nullptr, &outLen))
        return "";

    std::string result(outLen, '\0');
    if (!CryptBinaryToStringA(data.data(), static_cast<DWORD>(data.size()),
        CRYPT_STRING_BASE64 | CRYPT_STRING_NOCRLF, &result[0], &outLen))
        return "";

    // CryptBinaryToStringA includes a null terminator in outLen — trim it.
    if (!result.empty() && result.back() == '\0')
        result.pop_back();

    return result;
}

static std::vector<unsigned char> base64Decode(const std::string& text)
{
    DWORD outLen = 0;
    if (!CryptStringToBinaryA(text.c_str(), static_cast<DWORD>(text.size()),
        CRYPT_STRING_BASE64, nullptr, &outLen, nullptr, nullptr))
        return {};

    std::vector<unsigned char> result(outLen);
    if (!CryptStringToBinaryA(text.c_str(), static_cast<DWORD>(text.size()),
        CRYPT_STRING_BASE64, result.data(), &outLen, nullptr, nullptr))
        return {};

    return result;
}

// -------------------------------------------------------
// MachineGuid — HKLM\SOFTWARE\Microsoft\Cryptography\MachineGuid
// -------------------------------------------------------
std::string FDACrypto::getMachineGuid()
{
    char buf[64] = {};
    DWORD bufSize = sizeof(buf);
    DWORD type = REG_SZ;

    LONG res = RegGetValueA(
        HKEY_LOCAL_MACHINE,
        "SOFTWARE\\Microsoft\\Cryptography",
        "MachineGuid",
        RRF_RT_REG_SZ,
        &type,
        buf,
        &bufSize
    );

    if (res != ERROR_SUCCESS)
    {
        Logger::error("[CRYPTO] Failed to read MachineGuid, error : " + std::to_string(res));
        return "";
    }

    return std::string(buf);
}

// -------------------------------------------------------
// Key derivation — HMAC-SHA256(key=STATIC_SECRET, message=MachineGuid)
// -------------------------------------------------------
std::vector<unsigned char> FDACrypto::deriveKey()
{
    std::string machineGuid = getMachineGuid();
    if (machineGuid.empty())
        return {};

    std::vector<unsigned char> keyOut(32);

    BCRYPT_ALG_HANDLE hAlg = nullptr;
    BCRYPT_HASH_HANDLE hHash = nullptr;
    NTSTATUS status;

    status = BCryptOpenAlgorithmProvider(&hAlg, BCRYPT_SHA256_ALGORITHM, nullptr, BCRYPT_ALG_HANDLE_HMAC_FLAG);
    if (status < 0) { Logger::error("[CRYPTO] BCryptOpenAlgorithmProvider failed"); return {}; }

    status = BCryptCreateHash(hAlg, &hHash, nullptr, 0,
        const_cast<unsigned char*>(STATIC_SECRET), sizeof(STATIC_SECRET), 0);
    if (status < 0) { BCryptCloseAlgorithmProvider(hAlg, 0); Logger::error("[CRYPTO] BCryptCreateHash failed"); return {}; }

    status = BCryptHashData(hHash,
        reinterpret_cast<PUCHAR>(const_cast<char*>(machineGuid.data())),
        static_cast<ULONG>(machineGuid.size()), 0);
    if (status < 0) { BCryptDestroyHash(hHash); BCryptCloseAlgorithmProvider(hAlg, 0); Logger::error("[CRYPTO] BCryptHashData failed"); return {}; }

    status = BCryptFinishHash(hHash, keyOut.data(), static_cast<ULONG>(keyOut.size()), 0);

    BCryptDestroyHash(hHash);
    BCryptCloseAlgorithmProvider(hAlg, 0);

    if (status < 0) { Logger::error("[CRYPTO] BCryptFinishHash failed"); return {}; }

    return keyOut;
}

// -------------------------------------------------------
// encrypt() — AES-256-GCM, random 12-byte IV, output = iv || ciphertext || tag
// -------------------------------------------------------
std::string FDACrypto::encrypt(const std::string& plaintext)
{
    std::vector<unsigned char> key = deriveKey();
    if (key.empty()) return "";

    BCRYPT_ALG_HANDLE hAlg = nullptr;
    BCRYPT_KEY_HANDLE hKey = nullptr;
    NTSTATUS status;

    status = BCryptOpenAlgorithmProvider(&hAlg, BCRYPT_AES_ALGORITHM, nullptr, 0);
    if (status < 0) { Logger::error("[CRYPTO] encrypt: OpenAlgorithmProvider failed"); return ""; }

    BCryptSetProperty(hAlg, BCRYPT_CHAINING_MODE,
        reinterpret_cast<PUCHAR>(const_cast<wchar_t*>(BCRYPT_CHAIN_MODE_GCM)),
        sizeof(BCRYPT_CHAIN_MODE_GCM), 0);

    status = BCryptGenerateSymmetricKey(hAlg, &hKey, nullptr, 0, key.data(), static_cast<ULONG>(key.size()), 0);
    if (status < 0) { BCryptCloseAlgorithmProvider(hAlg, 0); Logger::error("[CRYPTO] encrypt: GenerateSymmetricKey failed"); return ""; }

    unsigned char iv[12];
    if (BCryptGenRandom(nullptr, iv, sizeof(iv), BCRYPT_USE_SYSTEM_PREFERRED_RNG) < 0)
    {
        BCryptDestroyKey(hKey); BCryptCloseAlgorithmProvider(hAlg, 0);
        Logger::error("[CRYPTO] encrypt: GenRandom (IV) failed");
        return "";
    }

    unsigned char tag[16];
    std::vector<unsigned char> ciphertext(plaintext.size());
    ULONG resultLen = 0;

    BCRYPT_AUTHENTICATED_CIPHER_MODE_INFO authInfo;
    BCRYPT_INIT_AUTH_MODE_INFO(authInfo);
    authInfo.pbNonce = iv;
    authInfo.cbNonce = sizeof(iv);
    authInfo.pbTag = tag;
    authInfo.cbTag = sizeof(tag);

    status = BCryptEncrypt(hKey,
        reinterpret_cast<PUCHAR>(const_cast<char*>(plaintext.data())), static_cast<ULONG>(plaintext.size()),
        &authInfo,
        nullptr, 0,
        ciphertext.empty() ? nullptr : ciphertext.data(), static_cast<ULONG>(ciphertext.size()),
        &resultLen, 0);

    BCryptDestroyKey(hKey);
    BCryptCloseAlgorithmProvider(hAlg, 0);

    if (status < 0) { Logger::error("[CRYPTO] encrypt: BCryptEncrypt failed"); return ""; }

    std::vector<unsigned char> combined;
    combined.reserve(sizeof(iv) + resultLen + sizeof(tag));
    combined.insert(combined.end(), iv, iv + sizeof(iv));
    combined.insert(combined.end(), ciphertext.begin(), ciphertext.begin() + resultLen);
    combined.insert(combined.end(), tag, tag + sizeof(tag));

    return base64Encode(combined);
}

// -------------------------------------------------------
// decrypt() — reverses encrypt(); empty string on any failure (including
// auth tag mismatch, which GCM uses to detect tampering/corruption)
// -------------------------------------------------------
std::string FDACrypto::decrypt(const std::string& base64Ciphertext)
{
    std::vector<unsigned char> key = deriveKey();
    if (key.empty()) return "";

    std::vector<unsigned char> combined = base64Decode(base64Ciphertext);
    if (combined.size() < 12 + 16) return ""; // too short to be iv+tag at minimum

    const unsigned char* iv = combined.data();
    const unsigned char* tag = combined.data() + combined.size() - 16;
    const unsigned char* ct = combined.data() + 12;
    size_t ctLen = combined.size() - 12 - 16;

    BCRYPT_ALG_HANDLE hAlg = nullptr;
    BCRYPT_KEY_HANDLE hKey = nullptr;
    NTSTATUS status;

    status = BCryptOpenAlgorithmProvider(&hAlg, BCRYPT_AES_ALGORITHM, nullptr, 0);
    if (status < 0) { Logger::error("[CRYPTO] decrypt: OpenAlgorithmProvider failed"); return ""; }

    BCryptSetProperty(hAlg, BCRYPT_CHAINING_MODE,
        reinterpret_cast<PUCHAR>(const_cast<wchar_t*>(BCRYPT_CHAIN_MODE_GCM)),
        sizeof(BCRYPT_CHAIN_MODE_GCM), 0);

    status = BCryptGenerateSymmetricKey(hAlg, &hKey, nullptr, 0, key.data(), static_cast<ULONG>(key.size()), 0);
    if (status < 0) { BCryptCloseAlgorithmProvider(hAlg, 0); Logger::error("[CRYPTO] decrypt: GenerateSymmetricKey failed"); return ""; }

    std::vector<unsigned char> plaintext(ctLen);
    ULONG resultLen = 0;

    BCRYPT_AUTHENTICATED_CIPHER_MODE_INFO authInfo;
    BCRYPT_INIT_AUTH_MODE_INFO(authInfo);
    authInfo.pbNonce = const_cast<PUCHAR>(iv);
    authInfo.cbNonce = 12;
    authInfo.pbTag = const_cast<PUCHAR>(tag);
    authInfo.cbTag = 16;

    status = BCryptDecrypt(hKey,
        const_cast<PUCHAR>(ct), static_cast<ULONG>(ctLen),
        &authInfo,
        nullptr, 0,
        plaintext.empty() ? nullptr : plaintext.data(), static_cast<ULONG>(plaintext.size()),
        &resultLen, 0);

    BCryptDestroyKey(hKey);
    BCryptCloseAlgorithmProvider(hAlg, 0);

    if (status < 0)
    {
        // Includes auth tag verification failure — tampered or corrupted value
        Logger::error("[CRYPTO] decrypt failed — value may be corrupted or tampered with");
        return "";
    }

    return std::string(reinterpret_cast<char*>(plaintext.data()), resultLen);
}